core/src/impl/Kokkos_Profiling_C_Interface.h

```diff
@@ -0,0 +1,97 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_PROFILING_C_INTERFACE_HPP
+#define KOKKOS_PROFILING_C_INTERFACE_HPP
+
+#ifdef __cplusplus
+#include <cstddef>
+#include <cstdint>
+#else
+#include <stddef.h>
+#include <stdint.h>
+#endif
+
+#define KOKKOSP_INTERFACE_VERSION 20191080
+
+struct Kokkos_Profiling_KokkosPDeviceInfo {
+  size_t deviceID;
+};
+
+struct Kokkos_Profiling_SpaceHandle {
+  char name[64];
+};
+
```

Comment:
Should we expect these two structs to change representation in the future?  I would suggest that each of them have two C interface functions provided given the current representation:

* an accessor for the current single field
* an equality comparison function

Either these are a point of indirection where the implementation can change, or they're legacy wrappers for the single field therein. If they're a point of indirection, a tool should be written to a defined, limited interface around these structs. If they're a legacy alias for a built-in type and it's safe to code that way, it will be easier and safer for tools to have access to the built-in type directly.

Reply:
@wrwilliams , in the case of SpaceHandle, that's certainly a legacy struct (though not one I'm in any rush to replace). But to your point more generally, I had been thinking of these headers are purely defining structs. A great idea in C++, but your comment reminds me it's kind of a bad one in C, in which abstraction really does tend to come from free function interfaces.

For this specific struct I'll likely leave it a legacy interface, but I'll keep your thinking in mind going forward.

Reply:
I mean, with an explicitly separated C interface header, you've got a fair bit of flexibility here to provide a C-only free function interface and as much/little encapsulation as you care to. If the C++ tools community finds it useful, great; if not, they can scorn it. But anyone stuck working in C (see many previous Tools Workshop gripe sessions, but it's not likely to change anytime soon) can't rely on default-generated operators, so either every C tool writes the part of that interface it needs or Kokkos writes an authoritative version of the ones that are allowed. I think you know my preference. :)

Reply:
Sorry I forgot to reply here. I don't think you're wrong, I think for our non-legacy types we should have authoritative interfaces to manipulate the types. Once I know this is going to be merged, my next step is probably to look at the Tuning interface and providing such functions